app/src/main/java/unipd/se18/ocrcamera/CameraActivity.java

```diff
@@ -0,0 +1,132 @@
+ *  so we can support full range of Android devices (API 15 and higher is 100% demographic)
+ *  I used this library that interfaces with the Camera (1) API to make the program
+ *      https://github.com/natario1/CameraView
+ *  working with master build of the library
+ *  The documentation starts here
+ *      https://github.com/natario1/CameraView#usage
+ */
+
+package com.example.elia.compatibilitycamera;
+
+import android.content.Intent;
+import android.os.Bundle;
+import android.support.v7.app.AppCompatActivity;
+import android.view.View;
+import android.content.pm.PackageManager;
+import android.support.annotation.NonNull;
+import android.widget.ImageButton;
+
+import com.otaliastudios.cameraview.CameraListener;
+import com.otaliastudios.cameraview.CameraView;
+import com.otaliastudios.cameraview.Gesture;
+import com.otaliastudios.cameraview.GestureAction;
+
+/**
+ * Activity for camera preview and taking photos
+ */
+public class CameraActivity extends AppCompatActivity {
+
+    //variable used to interface with the CameraView library
+    private CameraView camera;
+
+    //variable used to connect to the capturePhoto button
+    private ImageButton buttonPhoto;
+
+    /**
+     * onCreate method of the Android Activity Lifecycle
+     * @param savedInstanceState the bundle of the last instance state saved
+     * @author Bedin Elia
+     */
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_camera);
+        camera = findViewById(R.id.viewCamera);
+
+        //library method for automatic onResume, onPause, onDestroy
+        camera.setLifecycleOwner(this);
+
+        //Pinch to zoom support
+        camera.mapGesture(Gesture.PINCH, GestureAction.ZOOM);
+        //Tap to focus support
+        camera.mapGesture(Gesture.TAP, GestureAction.FOCUS_WITH_MARKER);
+        //Long tap to capture support (currently disabled, should be .CAPTURE)
+        camera.mapGesture(Gesture.LONG_TAP, GestureAction.NONE);
+
+        camera.addCameraListener(new CameraListener() {
+            /**
+             * Listening for photo taken by the camera
+             * @param picture the taken photo as an array of byte (a stream of data)
+             * @author Bedin Elia
+             */
+            public void onPictureTaken(byte[] picture) { onCapture(picture); }
+        });
+
+        //connecting the variable to the button capturePhoto
+        buttonPhoto = findViewById(R.id.capturePhoto);
+        //enabling the button
+        buttonPhoto.setEnabled(true);
+
+        buttonPhoto.setOnClickListener(new View.OnClickListener() {
+            /**
+             * Listening for button click, then call capturePhoto() to take a photo
+             * @param view the layout components
+             * @author Bedin Elia
+             */
+            @Override
+            public void onClick(View view) {
+                capturePhoto();
```

Comment:
In my opinion, it's clearer if the listener don't check anything, just call the method that's supposed to do the action we are listening for.
And to be clear, I'm not saying your proposal is wrong, quite the opposite.
What do you think? Is it so critical that I should move the check inside the listener?

Reply:
Rethinking about it with `buttonPhoto.setEnabled(false);` in the capturePhoto, the onClick of the listener and the capturePhoto are not launched until the `buttonPhoto.setEnabled(true);`, this behaviour is analogues to put that statement in the onClick of the listener, so it's not needed to put it in the listener, it works in both places. This is resolved!
I've written about removing also 91-92 because if the button is disabled the method capturePhoto is not launched actually. Could be `if (!buttonPhoto.isEnabled()) return;` of the line 92 useful now or in the future as a preventive checks or not in your opinion?

Reply:
Actually, I forgot to take out the if statement from the old implementation, you are right.
It has the `!buttonPhoto.isEnabled()` because I did a _find and replace all_.
I'm going to take it out now thank you very much.